File: medium/1358_number_of_substrings_containing_all_three_characters/solution.cpp

#include <string>

class Solution {
public:
  int numberOfSubstrings(std::string s) {
    int lastAbcMap[3] = {-1, -1, -1};
    int substringsCount = 0;

    for (int i = 0; i < static_cast<int>(s.length()); ++i) {
      lastAbcMap[s[i] - 'a'] = i;
      substringsCount +=
          std::min(lastAbcMap[0], std::min(lastAbcMap[1], lastAbcMap[2])) + 1;
    }

    return substringsCount;
  }
};

File: medium/1358_number_of_substrings_containing_all_three_characters/solution.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        lastAbcMap = [-1] * 3
        substringsCount = 0

        for i, c in enumerate(s):
            lastAbcMap[ord(c) - 97] = i
            substringsCount += min(lastAbcMap) + 1

        return substringsCount
